Scientist4JCore/src/main/java/com/github/rawls238/scientist4j/Experiment.java

```diff
@@ -59,77 +62,106 @@ public String getName() {
     }
 
     public T run(Supplier<T> control, Supplier<T> candidate) throws Exception {
-      Observation<T> controlObservation;
-      Optional<Observation> candidateObservation = Optional.empty();
-      if (Math.random() < 0.5) {
-        controlObservation = executeResult("control", controlTimer, control, true);
-        if (runIf() && enabled()) {
-          candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+        Observation<T> controlObservation;
+        Optional<Observation> candidateObservation = Optional.empty();
+        if (isAsync()) {
+            return runAsync(control, candidate);
+        } else {
+            if (Math.random() < 0.5) {
+                controlObservation = executeResult("control", controlTimer, control, true);
+                if (runIf() && enabled()) {
+                    candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+                }
+            } else {
+                if (runIf() && enabled()) {
+                    candidateObservation = Optional.of(executeResult("candidate", candidateTimer, candidate, false));
+                }
+                controlObservation = executeResult("control", controlTimer, control, true);
+            }
+            Result<T> result = new Result(this, controlObservation, candidateObservation, context);
+            publish(result);
+            return controlObservation.getValue();
         }
-      } else {
+    }
+
+    public T runAsync(Supplier<T> control, Supplier<T> candidate) throws Exception {
+        Future<Observation> observationFutureCandidate = null;
```

Comment:
Probably.  It does evaluate to an optional after running.  Changing the code would be this:

``` java
public T runAsync(Supplier<T> control, Supplier<T> candidate) throws Exception {
        Future<Optional<Observation>> observationFutureCandidate = null;
        Future<Observation<T>> observationFutureControl = executor.submit(() -> executeResult("control", controlTimer, control, true));
        if (runIf() && enabled()) {
            observationFutureCandidate = executor.submit(() -> Optional.of(executeResult("candidate", candidateTimer, candidate, false)));
        }

        Observation<T> controlObservation;
        try {
            controlObservation = observationFutureControl.get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
        Optional<Observation> candidateObservation = Optional.empty();
        if (observationFutureCandidate != null) {
            candidateObservation = observationFutureCandidate.get();
        }

        Result<T> result = new Result(this, controlObservation, candidateObservation, context);
        publish(result);
        return controlObservation.getValue();
    }
```

Are you concerned that the evaluation of the Optional.of outside the thread would add time in the current variation?